### DISCUSSION TITLE
Composition (transforming) of two vector fields

### DISCUSSION THREAD

**Discussion by ankitajoshi15** [voxelmorph/voxelmorph#215] [Q&A] [answered]
**Task** (what are you trying to do/register?)
Just trying to understand how two velocity fields are being composed, or transformed.

I understand how interpolation takes place either 'linear' or 'nearest' when one input is an image, and the other is a displacement field.

However, during integration when  "transform(vec,vec)" - function is called - which  essentially calls 'interpn' - how does the interpolation take place, since both the velocity fields will have x,y co ordinates? (or will have 2 nb_dims)

The description of the interpn function says this - 

> vol can have more dimensions than loc[i], in which case loc[i] acts as a slice 
    for the first dimensions

Please could you elaborate what this means?

**Comment by adalca** [comment, marked as answer]
@AnkitaJo ,
I'm not 100% sure I understand the question, but I'll try to explain

Let's forget for a second about the interpolation scheme.

When calling transform with an image and a vector, we basically say for each position: at this position (e.g. at pos [4, 5, 7]), get the intensity of the moving image at the location given by the displacement vector at this position. So loosely speaking, `moved[pos] = moving[pos + disp[pos]]` 

If the moving image is not an image, but in fact the field itself, the same holds:
`moved_field[pos] = field[pos + disp[pos]]` 

what does this field? well, as position `pos` we now have the vector field that was at the tip of disp[pos] vector. Therefore, *adding* these two: `moved_field[pos] + field[pos]` will now give us the vector addition of the two fields, which is what we want (it's the "composition"). 

So essentially there's no difference between how we treat vector fields and images (except that vector fields have nD channels)

The only problem with all of this is that when we do something like `field[pos + disp[pos]]`, that location might not be an integer, so we interpolate. But in some ways that's just a detail.

Does this make sense?

**Comment by ankitajoshi15** [comment]
@adalca Thank you so much for your response, that was precisely the explanation I needed to get a sense of how that function works. So basically, and please correct me if I am wrong,
For example:
A 2D vector field (vec) of say [X,Y,2], to get to new_vec[x,y,2] the function `transform(vec,vec)` will basically,

1. get moved positions: 
    new_vec = meshgrid +  old_vec (I think this will be the same as `moved_field[pos] + field[pos]`)
2. These values, as you mentioned, will mostly not be integers, for which we require interpolation, to actually pick the right cordinates from the old vec.

Am I on the right track here? Thanks a lot for your help!

**Comment by adalca** [comment]
@AnkitaJo 

Hmm, not quite, if i understand your description correctly.

assuming `old_vec` is just the **displacement** field (e.g. one that comes out of `voxelmorph`) which tells you the difference between each location and the one you're moving from, then   
`new_vec = mshgrid + old_vec`   
is just the **deformation** field (i.e. for each location, it tells you the **absolute** location you're moving from)

so if at location `[5, 3]` you have displacement `old_vec[5,3,:] == [1, 1]`, then `new_vec[5, 3]` will be `[6, 4]`. 

So this isn't quite the composition operation, it's just setting up the deformation field.

Now, once you've set this up, you actually have to **get the values** according to this field. So if you were warping an image, you'd do `new_image = old_image[new_vec]` (so that at the `new_image` location of `[5, 3]` you put whatever is located at `old_image` `[6, 4]`. This *warps* `old_image` to be `new_image`. The same with warping `old_field` to `new_field`, etc. and **that** is the operation i referred to above as `moved_field[pos] = field[pos + disp[pos]]`
